design/pool-configuration.md

```diff
@@ -0,0 +1,386 @@
+with a running speaker and an active endpoint for the service.
+
+In order to be able to announce L2 from a subset of nodes, this logic must be changed, checking for active endpoints only if the TrafficPolicy is local (which is handled by the [BGP implementation](https://github.com/metallb/metallb/blob/5d236989e61683daf1eaad00777b81ce7dd98c5d/speaker/bgp_controller.go#L153)).
+
+Besides that, implementing the address-pool / node association will be only reflected in a change in the `ShouldAnnounce` logic, eventually extending its signature with all the information needed to understand if the speaker must announce the ip or not.
+
+### BGP Peer Selector
+
+In order to be able to select a specific set of peers for a given pool we need to:
+
+* add a `name` field to the `Peer` structure so it can be selected by an AddressPool
+* add an optional list of bgp-peers, each one with its own (optional) node selector
+
+A configuration with bgp peer selection only would look like:
+
+```yaml
+    peers:
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      name: foo
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      name: bar
+    address-pools:
+    - name: default
+      protocol: bgp
+      addresses:
+      - 198.51.100.0/24
+      bgp-peers:
+        - name: foo
+        - name: bar
+```
+
+A configuration where bgp selection and nodes allocation would look like:
+
+```yaml
+    peers:
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      name: foo
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      name: bar
+    address-pools:
+    - name: default
+      protocol: bgp
+      addresses:
+      - 198.51.100.0/24
+      bgp-peers:
+        - name: foo
+        nodeSelector:
+          matchLabels:
+            network: first
+        - name: bar
+        nodeSelector:
+          matchLabels:
+            network: first
+```
+
+#### Implementation
+
+A service using an ip from a pool with a list of bgp-peers will not advertise the ip to peers outside that list.
+
+The BGP controller already contains a mapping between services and their BGP advertisement parameters taken from the address pool configuration.
+
+The structure is being filled [here](https://github.com/metallb/metallb/blob/5d236989e61683daf1eaad00777b81ce7dd98c5d/speaker/bgp_controller.go#L218).
+
+The idea here is to extend that with the list of peers that ip must be advertised to. In case the peer comes with a node selector, extra logic on the current node and the node selector will be in place to understand if to put the peer in the list of not.
+
+### Pool / Service - Namespace association
+
+The association of the pool with a particular service is orthogonal to the previously described feature.
```

Comment:
I am trying to convince myself of this statement :D. So, if I understand this correctly, if I assign an address pool to a particular service, we *can* reach that service from any node that has a backend for the service - this is regardless of what MetalLB does? .. which would seem to make them independent.

Reply:
This section is only about the way we associate the pool to a given service. Today is just random. The first pool that has free ips is associated to a given service [here](https://github.com/metallb/metallb/blob/997a177a762c6c8215ca9438083545529b675376/internal/allocator/allocator.go#L274).

Once we have an ip associated to the service, MetalLB will advertise it from a node that:
- has a speaker running (L2 only)
- in case of local traffic policy, has a running backend
- in case the node selector is implemented, the node must match the selector

This phase comes after the pool allocation to the service. 
The only relation I can see is that if a user cares about pinning a particular pool to a service (or to a namespace), the reason is likely because they want to leverage some of the extra tweakings that we are proposing here.